src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionSelector.cs

```diff
@@ -441,7 +441,16 @@ public int GetHashCode(string[] obj)
                 var hash = new HashCodeCombiner();
                 for (var i = 0; i < obj.Length; i++)
                 {
-                    hash.Add(obj[i], _valueComparer);
+                    var o = obj[i];
+
+                    // We want to match successfully in following cases:
+                    // Action has a required value of 'null' for 'area' and current request has area="" or area=null
+                    // Action has a required value of "" for 'foo' and current request has foo="" or foo=null
+                    if (string.IsNullOrEmpty(o))
+                    {
+                        o = null;
+                    }
+                    hash.Add(o, _valueComparer);
```

Comment:
I think you can explain what's going on here much more simply. Route values define `null` and `""` to be equivalent.